geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java

```diff
@@ -1881,9 +1881,8 @@ public void handleManagerDeparture(InternalDistributedMember theId, boolean memb
     removeHostedLocators(theId);
     redundancyZones.remove(theId);
 
-    if (theId.getVmKind() != ClusterDistributionManager.LOCATOR_DM_TYPE) {
-      stats.incNodes(-1);
-    }
+    stats.incNodes(-1);
+
```

Comment:
I think you're trying to change "nodes" to show the number of "servers" in the cluster but it's a different concept. "nodes" is a statistic in DistributionStats that reflects the number of peer-to-peer processes in the cluster, and we really need this statistic to remain as is.

I recommend adding one or more new attribute(s) to MemberMXBean or DistributedSystemMXBean which reflect the number of "non-locator members" or "servers". There are several related operations on DistributedSystemMXBean such listLocators(), listMembers(), listServers(). All three of these are "nodes".

listMembers() should show you all of the locators and servers (ie, any peer-to-peer process) in the cluster.

listLocators() should should you all of the locators only.

listServers() should show you all of the servers only.

There is a subtle difference between Attributes and Operations in JMX. You probably want an Attribute in this case so that it can be monitored with JMX APIs. But I'll leave further planning to you.

Reply:
I don't want to change "nodes" to show the number of "servers".
How it works by now?
If we have 2 locators and 2 servers the VisibleNodes attribute has the following values:
```
Locator 1: 2 (count Server 1 and Server 2)
Locator 2: 2 (count Server 1 and Server 2)
Server 1: 3 (count Server 2, and twice Server 1)
Server 2: 3 (count  Server 1, and twice Server 2)
```

With my changes it will count every member without itself:
```
Locator 1: 3 (count Locator 2, Server 1, and Server 2)
Locator 2: 3 (count Locator 1, Server 1, and Server 2)
Server 1: 3 (count Locator 1, Locator 2, and Server 2)
Server 2: 3 (count Locator 1, Locator 2, and Server 1)
```
What do you mean by the number of "peer-to-peer" processes in the cluster? 
Is that include both locators and servers?

Reply:
From the original code, it looks like it doesn't think a locator should be included in the count, so, shouldn't the correct count should be:

```
Locator 1: 2 (count Server 1 and Server 2) 
Locator 2: 2 (count Server 1 and Server 2)
Server 1: 2 (count Server 1 and Server 2) 
Server 2: 2 (count Server 1 and Server 2)
```
instead?

Reply:
In the description of this attribute state the following:
`The current number of nodes in this distributed system visible to this member.`

What are nodes there, both locators and servers or just servers?
Also, should the current server be visible to itself?

If we go with @jinmeiliao proposal we should change the description to be more clear.
`The current number of servers in this distributed system visible to this member.`

Reply:
I understand the purpose of the PR better now. However, the PR is still trying to redefine the "nodes" statistic in DistributionStats which is not correct. You should instead add a new attribute to MemberMXBean that returns the count of OTHER members in the cluster.

Here is the definition of the "nodes" statistic:
```
    final String nodesDesc = "The current number of nodes in this distributed system.";
```
So, if you have 2 servers and 2 locators, "nodes" should be 4 within all 4 JVMs.

The earliest version of GemFire had locators which were not members (ie not nodes), but locators now are definitely members (ie nodes). The next type of locators in GemFire were members (ie nodes) but without caches. As of GemFire 7.0, all locators have a cache. As of Geode 1.0, all locators are members with caches. The methods to create that earliest type of locator still exists but are deprecated... so deprecated internal methods is now the only way to create the thin locators that are not peer-to-peer members of the cluster.

Changing the attribute of an MXBean is _different_ from changing the meaning of the statistic. The statistic is not supposed to ignore locators.

Reply:
I'm marking this as resolved because I think developers who have worked on geode-membership should review it instead of me.